keps/sig-node/4381-dra-semantic-parameters/README.md

```diff
@@ -0,0 +1,1477 @@
+and standardization by Kubernetes. This could be used to describe topology
+across resources from different vendors, but this is out-of-scope for now.
+
+If a driver needs to reduce resource capacity, then there is a risk that a
+claim gets allocated using that capacity while the kubelet is updating a
+NodeResourceSlice. The implementations of semantic models must handle scenarios
+where more resources are allocated than available. The kubelet plugin of a DRA
+driver ***must*** double-check that the allocated resources are still available
+when NodePrepareResource is called. If not, the pod cannot start until the
+resource comes back. Treating this as a fatal error during pod admission would
+allow us to delete the pod and trying again with a new one.
+
+### Using semantic parameters
+
+The following is an example CRD which the developer of the
+`cards.dra.example.com` DRA driver might define as a valid claim parameters
+object for requesting access to its GPUs:
+
+```yaml
+kind: CardParameters
+apiVersion: dra.example.com/v1alpha1
+metadata:
+  name: my-parameters
+  namespace: user-namespace
+  uid: foobar-uid
+...
+spec:
+  count: 2
+  minimumRuntimeVersion: v12.0.0
+  minimumMemory: 8Gi
+  # "sharing" is a configuration parameter that does not
+  # get translated into the selector below.
+  sharing:
+    strategy: TimeSliced
+```
+
+Note that all fields in this CRD can be fully validated since it is owned by
+the DRA driver itself. This includes value ranges that are specific to the
+underlying hardware. There's no risk of using invalid attribute names because
+only the fields shown here are valid.
+
+With this CRD in place, a DRA driver controller is able to convert instances of
+it into a generic, in-tree `ResourceClaimParameters` object that the scheduler
+is able to understand.
+
+For the example above, the converted object would look as follows:
+
+```yaml
+kind: ResourceClaimParameters
+apiVersion: resource.k8s.io/v1alpha2
+
+metadata:
+  # This cannot be the same as my-parameters because parameter objects with a different
+  # type might also use it. Instead, the original object gets linked to below.
+  name: someArbitraryName
+  namespace: user-namespace
+
+generatedFrom:
+  name: my-parameters
+  kind: CardParameters
+  apiGroup: dra.example.com
+  uid: foobar-uid
+
+vendorParameters:
+  # A vendor can put any kind of object here to pass the configuration
+  # parameters down to the kubelet plugin. In this case, the vendor
+  # driver simply copied the entire CR. It could also be some
+  # separate, smaller configuration type.
+  kind: CardParameters
+  apiVersion: dra.example.com/v1alpha1
+  metadata:
+    name: my-parameters
+    namespace: user-namespace
+    uid: foobar-uid
+  ...
+  spec:
+    count: 2
+    minimumRuntimeVersion: v12.0.0
+    minimumMemory: 8Gi
+    sharing:
+      strategy: TimeSliced
+
+requests:
+- driverName: cards.dra.example.com
+  partitionable:
+    required:
+    - name: selected-gpu-0
```

Comment:
Just to verify my understanding:

So there are two "selected-gpu-X" stanzas, because of the "count: 2". But the selector is the same for them, because that's how the CRD is structured - it just has a count and the criteria, and those are de-normalized into individual identical requests.

Reply:
See my [previous comment](https://github.com/kubernetes/enhancements/pull/4384/files/4a0d7cdfd0df74d7813a6b889110f0044eb8cfbf#r1482106867) about why I thought it might be important to have distinct named requests here. If that was deemed not necessary, such a request could be simplified to something like the following:

```
requests:
- driverName: gpu.resource.nvidia.com
  partitionable:
  - count: 2
     selector: |-
       versions["runtimeVersion"] >= "v12.0.0" && quantities["memory"] >= "8Gi"
```

Reply:
I added:
> Each entry here is a request for one resource.
> They happen to have the same selector for the example, but could also be different.

I find this flexibility useful.